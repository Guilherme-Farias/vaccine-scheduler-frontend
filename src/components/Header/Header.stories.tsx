import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import Header from '.';

export default {
  title: 'Header',
  component: Header,
  parameters: {
    backgrounds: {
      default: 'minha-vacina-minor',
    },
  },
} as ComponentMeta<typeof Header>;

const Template: ComponentStory<typeof Header> = args => <Header {...args} />;

export const Desktop = Template.bind({});

export const Mobile = Template.bind({});
Mobile.parameters = {
  viewport: {
    defaultViewport: 'mobile2',
  },
};
