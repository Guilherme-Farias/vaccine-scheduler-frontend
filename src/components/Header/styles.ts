import styled, { css } from 'styled-components/macro';
import { transparentize } from 'polished';
import { DialogOverlay, DialogContent } from '@reach/dialog';
import { NavLink as NavLinkRRD } from 'react-router-dom';

export const Container = styled.header`
  ${({ theme }) => css`
    background-color: ${theme.colors.minorBg};
  `}
`;

export const Content = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    justify-content: space-between;
    height: ${theme.spacings['9xl']};
    color: ${theme.colors.white};
  `}
`;

export const DesktopNavigation = styled.nav`
  ${({ theme }) => css`
    @media ${theme.queries.phoneAndSmaller} {
      display: none;
    }
  `}
`;

export const NavList = styled.ul`
  display: flex;
  list-style: none;
`;

export const NavListItem = styled.li`
  ${({ theme }) => css`
    & + li {
      margin-left: ${theme.spacings.sm};
    }
  `}
`;

export const NavLink = styled(NavLinkRRD)`
  ${({ theme }) => css`
    display: inline-block;

    position: relative;
    color: inherit;
    font-size: ${theme.font.sizes.lg};
    margin: 3px ${theme.spacings.sm} 0;
    text-decoration: none;
    text-align: center;

    &:hover,
    &:focus,
    &.active {
      &::after {
        content: '';
        position: absolute;
        left: 0;
        display: block;
        height: 3px;
        background-color: ${theme.colors.tertiary};
        animation: hoverAnimation 0.2s forwards;
      }

      @keyframes hoverAnimation {
        from {
          width: 0;
        }
        to {
          width: 100%;
        }
      }
    }
  `};
`;

// Mobile Only
export const MobileActions = styled.div`
  ${({ theme }) => css`
    display: none;
    @media ${theme.queries.phoneAndSmaller} {
      display: revert;
    }
  `}
`;

export const IconContainer = styled.button`
  ${({ theme }) => css`
    color: inherit;
    display: block;
    border: none;
    background: transparent;
    text-align: left;
    padding: ${theme.spacings.md};
    &:focus:not(:focus-visible) {
      outline: none;
    }
  `}
`;

export const MobileMenu = styled(DialogOverlay)`
  ${({ theme }) => css`
    display: flex;
    justify-content: flex-end;
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;

    background: ${transparentize(0.5, theme.colors.black)};
  `}
`;

export const CloseIconContainer = styled(IconContainer)`
  ${({ theme }) => css`
    position: absolute;
    top: ${theme.spacings.md};
    right: ${theme.spacings.md};
  `}
`;

export const MobileContent = styled(DialogContent)`
  ${({ theme }) => css`
    display: flex;
    flex-direction: column;
    height: 100%;
    width: 300px;
    padding: ${theme.spacings['3xl']};
    background: ${theme.colors.mainBg};
    animation: slideIn 300ms 150ms backwards;
    @keyframes slideIn {
      from {
        transform: translateX(100%);
      }
      to {
        transform: translateX(0%);
      }
    }
  `}
`;

export const MobileNavigation = styled.nav`
  display: flex;
  flex: 1;
`;

export const MobileNavList = styled.ul`
  ${({ theme }) => css`
    display: flex;
    flex: 1;
    flex-direction: column;
    list-style: none;
    justify-content: center;
    align-items: start;
    gap: ${theme.spacings['4xl']};
    > ${NavListItem} {
      margin-left: revert;
    }
  `}
`;
