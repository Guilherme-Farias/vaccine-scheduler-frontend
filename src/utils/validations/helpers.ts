import Joi from 'joi';
import { IParamErrors } from 'types/global';

export function getFieldErrors(objError: Joi.ValidationResult) {
  const errors: IParamErrors = {};

  if (objError.error) {
    objError.error.details.forEach(err => {
      errors[err.path.join('.')] = err.message;
    });
  }

  return errors;
}
