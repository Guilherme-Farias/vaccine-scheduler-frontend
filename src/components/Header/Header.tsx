import React from 'react';

import Logo from 'components/Logo';
import MaxWidthWrapper from 'components/MaxWidthWrapper';
import VisuallyHidden from 'components/VisuallyHidden';

import { FaBars as MenuIcon, FaTimes as CloseIcon } from 'react-icons/fa';
import * as S from './styles';

const Header: React.FC = () => {
  const [showMobileMenu, setShowMobileMenu] = React.useState(false);

  return (
    <S.Container>
      <MaxWidthWrapper>
        <S.Content>
          <S.NavLink to="/" end>
            <Logo />
          </S.NavLink>

          <S.DesktopNavigation>
            <S.NavList>
              <S.NavListItem>
                <S.NavLink
                  to="/agendamentos"
                  title="Página de listagem de agendamentos"
                  end
                >
                  Agendamentos
                </S.NavLink>
              </S.NavListItem>

              <S.NavListItem>
                <S.NavLink
                  to="/agendamentos/novo"
                  title="Página para criação de um novo agendamento"
                  end
                >
                  Agendar
                </S.NavLink>
              </S.NavListItem>
            </S.NavList>
          </S.DesktopNavigation>

          <S.MobileActions>
            <S.IconContainer onClick={() => setShowMobileMenu(true)}>
              <MenuIcon size={24} />
              <VisuallyHidden>Abrir menu</VisuallyHidden>
            </S.IconContainer>
          </S.MobileActions>

          <S.MobileMenu
            isOpen={showMobileMenu}
            onDismiss={() => setShowMobileMenu(false)}
          >
            <S.MobileContent aria-label="Menu de navegação">
              <S.CloseIconContainer onClick={() => setShowMobileMenu(false)}>
                <CloseIcon size={24} />
                <VisuallyHidden>Dispensar menu</VisuallyHidden>
              </S.CloseIconContainer>

              <S.MobileNavigation>
                <S.MobileNavList>
                  <S.NavListItem>
                    <S.NavLink
                      to="/"
                      title="Página de listagem de agendamentos"
                      end
                    >
                      Início
                    </S.NavLink>
                  </S.NavListItem>

                  <S.NavListItem>
                    <S.NavLink
                      to="/agendamentos"
                      title="Página de listagem de agendamentos"
                      end
                    >
                      Agendamentos
                    </S.NavLink>
                  </S.NavListItem>

                  <S.NavListItem>
                    <S.NavLink
                      to="/agendamentos/novo"
                      title="Página para criação de um novo agendamento"
                      end
                    >
                      Agendar
                    </S.NavLink>
                  </S.NavListItem>
                </S.MobileNavList>
              </S.MobileNavigation>
            </S.MobileContent>
          </S.MobileMenu>
        </S.Content>
      </MaxWidthWrapper>
    </S.Container>
  );
};
export default Header;
