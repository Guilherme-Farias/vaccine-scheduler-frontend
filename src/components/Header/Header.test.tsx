import { render, screen, userEvent, within } from 'utils/tests/helpers';

import Header from '.';

describe('<Header />', () => {
  it('should render the Header', () => {
    const { container } = render(<Header />);

    expect(screen.getByText(/abrir menu/i)).toBeInTheDocument();
    expect(
      screen.getByRole('link', { name: /^minha vacina$/i }),
    ).toHaveAttribute('href', '/');

    expect(
      screen.getByRole('link', {
        name: /^página de listagem de agendamentos$/i,
      }),
    ).toHaveAttribute('href', '/agendamentos');

    expect(
      screen.getByRole('link', {
        name: /^página para criação de um novo agendamento$/i,
      }),
    ).toHaveAttribute('href', '/agendamentos/novo');

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should handle the open/close mobile menu', async () => {
    render(<Header />);

    // Checks if the menu is closed
    expect(screen.queryByRole('dialog')).not.toBeInTheDocument();

    // Open the menu
    await userEvent.click(screen.getByText(/abrir menu/i));
    const mobileMenu = screen.getByRole('dialog');

    // Check if the menu is open
    expect(mobileMenu).toBeInTheDocument();
    expect(mobileMenu).toHaveAttribute('aria-label', 'Menu de navegação');

    // Close menu by close button
    const inMobileMenu = within(screen.getByRole('dialog'));
    await userEvent.click(
      inMobileMenu.getByRole('button', { name: /dispensar menu/i }),
    );

    // Checks if the menu is closed
    expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
  });

  it('should close the mobile menu when clicking on the overlay', async () => {
    render(<Header />);

    // Checks if the menu is closed
    expect(screen.queryByRole('dialog')).not.toBeInTheDocument();

    // Open the menu
    await userEvent.click(screen.getByText(/abrir menu/i));
    const mobileMenu = screen.getByRole('dialog');
    const overlay = mobileMenu.parentElement;

    // Check if the menu is open
    expect(mobileMenu).toBeInTheDocument();
    expect(mobileMenu).toHaveAttribute('aria-label', 'Menu de navegação');

    // Close menu on click overlay
    await userEvent.click(overlay!);

    // Checks if the menu is closed
    expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
  });

  it('should close the mobile menu with the Escape key', async () => {
    render(<Header />);

    // Checks if the menu is closed
    expect(screen.queryByRole('dialog')).not.toBeInTheDocument();

    // Open the menu
    await userEvent.click(screen.getByText(/abrir menu/i));
    const mobileMenu = screen.getByRole('dialog');

    // Check if the menu is open
    expect(mobileMenu).toBeInTheDocument();
    expect(mobileMenu).toHaveAttribute('aria-label', 'Menu de navegação');

    // Close the menu by pressing the Escape key
    await userEvent.keyboard('{Esc}');

    // Checks if the menu is closed
    expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
  });
});
