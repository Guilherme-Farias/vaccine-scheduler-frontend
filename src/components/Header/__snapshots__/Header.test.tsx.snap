// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should render the Header 1`] = `
.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #FFFFFF;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.875rem;
}

.c5 {
  margin-right: 0.25rem;
}

.c5 > svg {
  font-size: inherit;
}

.c6 {
  font-size: inherit;
  font-weight: 400;
}

.c1 {
  width: 100%;
  max-width: 81.25rem;
  margin-left: auto;
  margin-right: auto;
  padding-left: calc(2rem / 2);
  padding-right: calc(2rem / 2);
}

.c12 {
  position: absolute;
  overflow: hidden;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  width: 1px;
  margin: -1px;
  padding: 0;
  border: 0;
}

.c0 {
  background-color: #D03E35;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  height: 5rem;
  color: #FFFFFF;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  list-style: none;
}

.c9 + li {
  margin-left: 0.75rem;
}

.c3 {
  display: inline-block;
  position: relative;
  color: inherit;
  font-size: 1.125rem;
  margin: 3px 0.75rem 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
}

.c3:hover::after,
.c3:focus::after,
.c3.active::after {
  content: '';
  position: absolute;
  left: 0;
  display: block;
  height: 3px;
  background-color: #FFC801;
  -webkit-animation: hoverAnimation 0.2s forwards;
  animation: hoverAnimation 0.2s forwards;
}

.c10 {
  display: none;
}

.c11 {
  color: inherit;
  display: block;
  border: none;
  background: transparent;
  text-align: left;
  padding: 1rem;
}

.c11:focus:not(:focus-visible) {
  outline: none;
}

@media (max-width:48rem) {
  .c4 {
    font-size: 1.5rem;
  }
}

@media (max-width:48rem) {
  .c7 {
    display: none;
  }
}

@media (max-width:48rem) {
  .c10 {
    display: revert;
  }
}

<header
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <a
        aria-current="page"
        class="c3 active"
        href="/"
      >
        <div
          class="c4"
        >
          <div
            class="c5"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M201.5 174.8l55.7 55.8c3.1 3.1 3.1 8.2 0 11.3l-11.3 11.3c-3.1 3.1-8.2 3.1-11.3 0l-55.7-55.8-45.3 45.3 55.8 55.8c3.1 3.1 3.1 8.2 0 11.3l-11.3 11.3c-3.1 3.1-8.2 3.1-11.3 0L111 265.2l-26.4 26.4c-17.3 17.3-25.6 41.1-23 65.4l7.1 63.6L2.3 487c-3.1 3.1-3.1 8.2 0 11.3l11.3 11.3c3.1 3.1 8.2 3.1 11.3 0l66.3-66.3 63.6 7.1c23.9 2.6 47.9-5.4 65.4-23l181.9-181.9-135.7-135.7-64.9 65zm308.2-93.3L430.5 2.3c-3.1-3.1-8.2-3.1-11.3 0l-11.3 11.3c-3.1 3.1-3.1 8.2 0 11.3l28.3 28.3-45.3 45.3-56.6-56.6-17-17c-3.1-3.1-8.2-3.1-11.3 0l-33.9 33.9c-3.1 3.1-3.1 8.2 0 11.3l17 17L424.8 223l17 17c3.1 3.1 8.2 3.1 11.3 0l33.9-34c3.1-3.1 3.1-8.2 0-11.3l-73.5-73.5 45.3-45.3 28.3 28.3c3.1 3.1 8.2 3.1 11.3 0l11.3-11.3c3.1-3.2 3.1-8.2 0-11.4z"
              />
            </svg>
          </div>
          <h1
            class="c6"
          >
            Minha 
            <strong>
              Vacina
            </strong>
          </h1>
        </div>
      </a>
      <nav
        class="c7"
      >
        <ul
          class="c8"
        >
          <li
            class="c9"
          >
            <a
              class="c3"
              href="/agendamentos"
              title="Página de listagem de agendamentos"
            >
              Agendamentos
            </a>
          </li>
          <li
            class="c9"
          >
            <a
              class="c3"
              href="/agendamentos/novo"
              title="Página para criação de um novo agendamento"
            >
              Agendar
            </a>
          </li>
        </ul>
      </nav>
      <div
        class="c10"
      >
        <button
          class="c11"
        >
          <svg
            fill="currentColor"
            height="24"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
            />
          </svg>
          <div
            class="c12"
          >
            Abrir menu
          </div>
        </button>
      </div>
    </div>
  </div>
</header>
`;
